File: go/source/scene.go
package source

import (
	"math/rand"
)

// Scene represents the view
type Scene struct {
	source, corner, horizon, vertical Vector
	list                              Hittable
	aperture                          float64
}

// NewScene returns a new View
func NewScene(source, corner, horizon, vertical Vector, aperture float64) Scene {
	return Scene{source, corner, horizon, vertical, nil, aperture}
}

// Source returns the source of the scene
func (scn Scene) Source() Vector { return scn.source }

// Corner returns the corner of the scene
func (scn Scene) Corner() Vector { return scn.corner }

// Horizon returns the horizon of the scene
func (scn Scene) Horizon() Vector { return scn.horizon }

// Vertical returns the vertical of the scene
func (scn Scene) Vertical() Vector { return scn.vertical }

// Aperture returns the aperture of the scene
func (scn Scene) Aperture() float64 { return scn.aperture }

// SourceTo sets the source of the scene
func (scn *Scene) SourceTo(source Vector) { scn.source = source }

// CornerTo sets the corner of the scene
func (scn *Scene) CornerTo(corner Vector) { scn.corner = corner }

// HorizonTo sets the horizon of the scene
func (scn *Scene) HorizonTo(horizon Vector) { scn.horizon = horizon }

// VerticalTo sets the vertical of the scene
func (scn *Scene) VerticalTo(vertical Vector) { scn.vertical = vertical }

// ApertureTo returns the aperture of the scene
func (scn *Scene) ApertureTo(aperture float64) { scn.aperture = aperture }

// Save switches the Scene's backend
func (scn *Scene) Save(h Hittable) {
	scn.list = h
}

// ColorTrace tracks the color of a path
func (scn Scene) ColorTrace(starting, towards Vector, depth int, gen *rand.Rand) Vector {
	color := VectorUniform(1.)
	for d := 0; d < depth; d++ {
		if data := scn.Hit(starting, towards); data.HasHit() {
			matter := data.Matter()
			reflected := matter.Scatter(towards, data.Normal(), gen)
			color.IMul(matter.Albedo())
			starting, towards = data.Point(), reflected
		} else {
			t := .5 * (towards.Unit().Y() + 1.)
			background := VectorUniform(1.).MulS(1. - t).Add(NewVector(.5, .7, 1.).MulS(t))
			return color.Mul(background)
		}
	}
	return VectorO()
}

// Color determines the color at a given position
func (scn Scene) Color(x, y, ns, depth int, dx, dy float64, gen *rand.Rand) (r, g, b uint8) {
	i, j := RandomDisk(scn.aperture, gen)
	h, v := scn.horizon.Unit().MulS(i), scn.vertical.Unit().MulS(j)
	start := scn.source.Add(h).Add(v)

	var color Vector
	for s := 0; s < ns; s++ {
		i, j := (float64(x)+gen.Float64())/dx, (float64(y)+gen.Float64())/dy
		end := scn.corner.Add(scn.horizon.MulS(i)).Add(scn.vertical.MulS(j))
		towards := end.Sub(start)

		color.IAdd(scn.ColorTrace(start, towards, depth, gen))
	}

	pixel := color.DivS(float64(ns)).MulS(255.999)
	r = uint8(pixel.X())
	g = uint8(pixel.Y())
	b = uint8(pixel.Z())
	return
}

// Hit implements Hittable for Scene
func (scn Scene) Hit(source, towards Vector) HitData {
	return scn.list.Hit(source, towards)
}

// Bounds implements Hittable for Scene
func (scn Scene) Bounds() Box {
	return scn.list.Bounds()
}

// RandomDisk creates a random pair of tuple that lies in a unit disk
func RandomDisk(radius float64, gen *rand.Rand) (x, y float64) {
	for {
		x, y = gen.Float64(), gen.Float64()
		if x*x+y*y <= 1 {
			return x * radius, y * radius
		}
	}
}

File: java/src/main/java/com/app/source/Scene.java
package com.app.source;

import java.util.concurrent.ThreadLocalRandom;
import java.util.stream.IntStream;

public final class Scene implements Hittable {
    private Vector source;
    private Vector corner;
    private Vector horizon;
    private Vector vertical;

    private Hittable list;
    private double aperture;

    public Scene(Vector source, Vector corner, Vector horizon, Vector vertical, double aperture) {
        this.source = source;
        this.corner = corner;
        this.horizon = horizon;
        this.vertical = vertical;
        this.list = null;
        this.aperture = aperture;
    }

    public Scene(Vector source, Vector corner, Vector horizon, Vector vertical, Hittable list, double aperture) {
        this.source = source;
        this.corner = corner;
        this.horizon = horizon;
        this.vertical = vertical;
        this.list = list;
        this.aperture = aperture;
    }

    public Vector source() {
        return this.source;
    }

    public void source(Vector v) {
        this.source = v;
    }

    public Vector corner() {
        return this.corner;
    }

    public void corner(Vector v) {
        this.corner = v;
    }

    public Vector horizon() {
        return this.horizon;
    }

    public void horizon(Vector v) {
        this.horizon = v;
    }

    public Vector vertical() {
        return this.vertical;
    }

    public void vertical(Vector v) {
        this.vertical = v;
    }

    public Hittable list() {
        return this.list;
    }

    public void list(Hittable a) {
        this.list = a;
    }

    public double aperture() {
        return this.aperture;
    }

    public void aperture(double a) {
        this.aperture = a;
    }

    public void save(Hittable h) {
        list = h;
    }

    public Vector color_trace(Vector starting, Vector towards, int depth) {
        var color = Vector.uniform(1.);
        for (int d = 0; d < depth; ++d) {
            var data = hit(starting, towards);

            if (data.isHit()) {
                var matter = data.matter();
                var reflected = matter.scatter(towards, data.normal());

                color.imul(matter.albedo());

                starting = data.point();
                towards = reflected;
            } else {
                double t = .5 * (towards.unit().y() + 1.);
                var back = Vector.uniform(1.).mul(1. - t).add(new Vector(.5, .7, 1.).mul(t));
                return color.mul(back);
            }
        }
        return Vector.o();
    }

    public int[] color(int x, int y, int ns, int depth, double dx, double dy) {
        var random = ThreadLocalRandom.current();

        double[] a = randomDisk(aperture);
        assert a.length == 2;
        double ai = a[0];
        double aj = a[1];

        Vector h = horizon.unit().mul(ai);
        Vector v = vertical.unit().mul(aj);
        Vector start = source.add(h).add(v);

        var color = IntStream.range(0, ns).sequential().mapToObj(_i -> {
            double i = ((double) x + random.nextDouble()) / dx;
            double j = ((double) y + random.nextDouble()) / dy;

            Vector end = corner.add(horizon.mul(i).add(vertical.mul(j)));
            Vector towards = end.sub(start);

            return color_trace(start, towards, depth);
        }).reduce(Vector.o(), Vector::add);

        var pixel = color.div(ns).mul(255.999);
        return new int[] { (int) pixel.x(), (int) pixel.y(), (int) pixel.z() };
    }

    public static double[] randomDisk(double radius) {
        var random = ThreadLocalRandom.current();
        for (;;) {
            double x = random.nextDouble();
            double y = random.nextDouble();

            if (x * x + y * y <= 1.) {
                return new double[] { x * radius, y * radius };
            }
        }
    }

    @Override
    public HitData hit(Vector source, Vector towards) {

        return list.hit(source, towards);
    }

    @Override
    public Box bounds() {
        return list.bounds();
    }
}
